File: leetcode/leetcode_cpp/construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

/*
https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal

Solution1. divide and conquer(recursion)
time: o(n^2) visit node(n) * find inroder pivot(n)
space: o(n)

Solution2. divide and conquer + map (find pivot index quick)
time: o(n)
space: o(n) recursion + map

*/


class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return buildTree_2(preorder, inorder);
    }

    TreeNode* buildTree_1(vector<int>& preorder, vector<int>& inorder) {
        int n = inorder.size();
        int idx = 0;
        return build(preorder, inorder, idx, 0, n-1);
    }

    TreeNode* build(vector<int>& preorder, vector<int>& inorder, int& idx, int left, int right) {
        if (left > right) return nullptr;
        int pivot = left;
        int val = preorder[idx];
        while (inorder[pivot] != val) pivot++; // improve this with map

        idx++;
        auto node = new TreeNode(val);
        node->left = build(preorder, inorder, idx, left, pivot-1);
        node->right = build(preorder, inorder, idx, pivot+1, right);
        return node;
    }

    map<int,int> mp;    
    TreeNode* buildTree_2(vector<int>& preorder, vector<int>& inorder) {
        for (int i=0; i<inorder.size(); ++i)
            mp[inorder[i]] = i;
        int idx = 0;
        return build_2(preorder, inorder, idx, 0, inorder.size() - 1);
    }

    TreeNode* build_2(vector<int>& preorder, vector<int>& inorder, int& idx, int left, int right) {
        if (left > right) return nullptr;
        int val = preorder[idx];
        int pivot = mp[val];
        
        idx++;
        auto node = new TreeNode(val);
        node->left = build_2(preorder, inorder, idx, left, pivot-1);
        node->right = build_2(preorder, inorder, idx, pivot+1, right);
        return node;
    }
};

File: leetcode/leetcode_py/construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

'''
https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal

Solution1. divide and conquer(recursion)
time: o(n^2) visit node(n) * find inroder pivot(n)
space: o(n)

Solution2. divide and conquer + map (find pivot index quick)
time: o(n)
space: o(n) recursion + map
'''

class Solution:
    def buildTree_1(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        def build(l, r):
            if l > r:
                return None
            val = preorder[self.idx]
            pivot = l
            while inorder[pivot] != val:
                pivot += 1

            self.idx += 1
            node = TreeNode(val)
            node.left = build(l, pivot-1)
            node.right = build(pivot+1, r)
            return node
        self.idx = 0
        return build(0, len(inorder) - 1)

    def buildTree_2(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        def build(l, r):
            if l > r:
                return None
            val = preorder[self.idx]
            pivot = m[val]

            self.idx += 1
            node = TreeNode(val)
            node.left = build(l, pivot-1)
            node.right = build(pivot+1, r)
            return node
        self.idx = 0
        m = {}
        for i, v in enumerate(inorder):
            m[v] = i
        return build(0, len(inorder) - 1)

    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        return self.buildTree_2(preorder, inorder)
            
        